Sources/StreamChat/Query/Filter+Predicate.swift

```diff
@@ -0,0 +1,88 @@
+//
+// Copyright © 2023 Stream.io Inc. All rights reserved.
+//
+
+import Foundation
+
+extension Filter {
+    var predicate: NSPredicate? {
+        guard let op = FilterOperator(rawValue: self.operator) else { return nil }
+
+        let unableToResolve: () -> NSPredicate? = {
+            let message = """
+            The filter passed to the query could not be resolved at runtime.
+            Please check the documentation at: https://getstream.io/chat/docs/sdk/ios/client/controllers/channels/#%EF%B8%8F-when-the-filter-cannot-be-auto-resolved
+            Otherwise you can have issues when multiples channel list controllers are alive at the same time
+            """
+            log.assertionFailure(message)
+            return nil
+        }
+
+        let dtoKey = autoKey?.dtoKey ?? key
+
+        let makeComparablePredicate: (_ compareToken: String) -> NSPredicate? = { compareToken in
+            guard let key = dtoKey else { return unableToResolve() }
+            let valueToUse: CVarArg = autoKey?.value ?? value.description
+            return NSPredicate(format: "\(key) \(compareToken) %@", valueToUse)
+        }
+
+        let makeCompoundPredicate: (_ logicalType: NSCompoundPredicate.LogicalType) -> NSPredicate? = { logicalType in
+            guard let value = self.value as? [Filter] else { return unableToResolve() }
+            let predicates = value.compactMap(\.predicate)
+            guard !predicates.isEmpty else { return unableToResolve() }
+            return NSCompoundPredicate(type: logicalType, subpredicates: predicates)
+        }
+
+        switch op {
+        case .equal:
+            return makeComparablePredicate("==")
+        case .notEqual:
+            return makeComparablePredicate("!=")
+        case .greater:
+            return makeComparablePredicate(">")
+        case .greaterOrEqual:
+            return makeComparablePredicate(">=")
+        case .less:
+            return makeComparablePredicate("<")
+        case .lessOrEqual:
+            return makeComparablePredicate("<=")
+        case .in:
+            guard let key = dtoKey else { return unableToResolve() }
+            guard let valueArray = value as? [FilterValue] else { return unableToResolve() }
+            return NSCompoundPredicate(
+                type: .and,
+                subpredicates: valueArray.map {
+                    NSPredicate(format: "ANY \(key) == %@", $0.description)
+                }
+            )
+        case .notIn:
+            guard let key = dtoKey else { return unableToResolve() }
+            guard let valueArray = value as? NSArray else { return unableToResolve() }
+            return NSCompoundPredicate(format: "NONE \(key) IN %@", valueArray)
+        case .contains:
+            return makeComparablePredicate("IN")
```

Comment:
I'm not sure this is quite true. The original implementation was this:
```swift
guard let key = dtoKey else { return nil }
guard let valueArray = value as? [CVarArg] else { return nil }
return NSPredicate(format: "\(key) IN %@", valueArray)
```

So the value needs to be `[CVarArg]`. Did you test if this produced the same result?

Reply:
Another reason why to choose `IN` is a potential performance benefit([reference](https://stackoverflow.com/a/3074731)). SQL seems that uses binary search for IN while for AND/OR statements will evaluate them one by one.